habitat_baselines/common/obs_transformers.py

```diff
@@ -217,113 +218,496 @@ def from_config(cls, config: Config):
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Rotate world points according to camera rotation
+        world_pts = self.worldcoord2camcoord(world_pts)
+
+        x, y, z = world_pts[..., 0], world_pts[..., 1], world_pts[..., 2]
+        # x,y,z to theta, phi
+        theta = torch.atan2(x, z)
+        c = torch.sqrt(x * x + z * z)
+        phi = torch.atan2(y, c)
+
+        # For grid_sample, -1 <= proj_pts <= 1
+        mapx = theta / np.pi
+        mapy = phi / (np.pi / 2)
+        proj_pts = torch.stack([mapx, mapy], dim=-1)
+
+        # All points in image are valid
+        valid_mask = torch.full(proj_pts.shape[:2], True, dtype=torch.bool)
+        return proj_pts, valid_mask
+
+    def unprojection(
+        self, with_rotation: bool = True
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        theta_map, phi_map = self.get_theta_phi_map(self.img_h, self.img_w)
+        unproj_pts = self.angle2sphere(theta_map, phi_map)
+        # All points in image are valid
+        valid_mask = torch.full(unproj_pts.shape[:2], True, dtype=torch.bool)
+        # Rotate unproj_pts points according to camera rotation
+        if with_rotation:
+            unproj_pts = self.camcoord2worldcoord(unproj_pts)
+        return unproj_pts, valid_mask
+
+    # Get theta and phi map
+    def get_theta_phi_map(self, img_h: int, img_w: int) -> torch.Tensor:
+        phi, theta = torch.meshgrid(torch.arange(img_h), torch.arange(img_w))
+        theta_map = (theta + 0.5) * 2 * np.pi / img_w - np.pi
+        phi_map = (phi + 0.5) * np.pi / img_h - np.pi / 2
+        return theta_map, phi_map
+
+    # Project on unit sphere
+    def angle2sphere(
+        self, theta_map: torch.Tensor, phi_map: torch.Tensor
+    ) -> torch.Tensor:
+        sin_theta = torch.sin(theta_map)
+        cos_theta = torch.cos(theta_map)
+        sin_phi = torch.sin(phi_map)
+        cos_phi = torch.cos(phi_map)
+        return torch.stack(
+            [cos_phi * sin_theta, sin_phi, cos_phi * cos_theta], dim=-1
+        )
+
+
+class FisheyeCamera(CameraProjection):
+    r"""This is the fisheye camera class.
+    The camera model is based on the Double Sphere Camera Model (Usenko et. al.;3DV 2018).
+    Paper: https://arxiv.org/abs/1807.08957
+    Implementation: https://github.com/matsuren/dscamera
+    """
+
+    def __init__(
+        self,
+        img_h: int,
+        img_w: int,
+        fish_fov: float,
+        cx: float,
+        cy: float,
+        fx: float,
+        fy: float,
+        xi: float,
+        alpha: float,
+        R: Optional[torch.Tensor] = None,
+    ):
+        """Args:
+        img_h: (int) the height of fisheye camera image
+        img_w: (int) the width of fisheye camera image
+        fish_fov: (float) the fov of fisheye camera in degrees
+        cx, cy: (float) the optical center of the fisheye camera
+        fx, fy, xi, alpha: (float) the fisheye camera model parameters
+        R: (torch.Tensor) 3x3 rotation matrix of camera
```

Comment:
Should make a note: in Radians

Reply:
You mean `fov_rad = self.fish_fov / 180 * np.pi`? I added a comment.

Reply:
I actually meant the rotation matrix. In general, angular units should always be defined IMO.

Reply:
Hmm? Is the rotation matrix without units? (neither radians nor degrees)